### PR TITLE
Master UI has link to Application UI

### PR DESCRIPTION
**Pull request by squito**
Its a pain to go through the logs to find the link to the Storage Dashboard, so now the master UI has a link to it.

Even though its just the BlockMangerUI right now, I called it Application UI in the code & the link, b/c I assume eventually we'll want it to serve more than just storage info.

### CODE REVIEW

File: core/src/main/scala/spark/deploy/ApplicationDescription.scala [hunk 1]
```diff
@@ -5,7 +5,9 @@ private[spark] class ApplicationDescription(
     val cores: Int,
     val memoryPerSlave: Int,
     val command: Command,
-    val sparkHome: String)
+    val sparkHome: String,
+    val appUIHost: String,
+    val appUIPort: Int)
```
Comment (markhamstra) [inline core/src/main/scala/spark/deploy/ApplicationDescription.scala]: This applies more broadly than to just your additions, but is there any reason to keep appUIHost and appUIPort as separate elements?  If they always need to go together as a consistent pair, I think I'd rather see them packaged as at least a tuple.

Comment (squito) [inline core/src/main/scala/spark/deploy/ApplicationDescription.scala]: yeah, this is a great point.  I wondered about this myself, but made it this way just for consistency.

Comment (mateiz) [inline core/src/main/scala/spark/deploy/ApplicationDescription.scala]: I agree on this; in fact it should just be an appUiUrl or something like that.

**Comment by squito** [comment]
updated to use combined `appUiUrl`

File: core/src/main/scala/spark/storage/BlockManagerUI.scala [hunk 2]
```diff
@@ -20,19 +20,20 @@ class BlockManagerUI(val actorSystem: ActorSystem, blockManagerMaster: ActorRef,
   val STATIC_RESOURCE_DIR = "spark/deploy/static"
 
   implicit val timeout = Duration.create(System.getProperty("spark.akka.askTimeout", "10").toLong, "seconds")
+  val host = Utils.localHostName()
+  val port = if (System.getProperty("spark.ui.port") != null) {
+    System.getProperty("spark.ui.port").toInt
+  } else {
+    // TODO: Unfortunately, it's not possible to pass port 0 to spray and figure out which
```
Comment (rxin) [inline core/src/main/scala/spark/storage/BlockManagerUI.scala:27]: What is the TODO item here?

**Comment by rxin** [comment]
Jenkins, test this please.

File: core/src/main/scala/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala [hunk 3]
```diff
@@ -31,7 +31,8 @@ private[spark] class SparkDeploySchedulerBackend(
     val command = Command("spark.executor.StandaloneExecutorBackend", args, sc.executorEnvs)
     val sparkHome = sc.getSparkHome().getOrElse(
       throw new IllegalArgumentException("must supply spark home for spark standalone"))
-    val appDesc = new ApplicationDescription(appName, maxCores, executorMemory, command, sparkHome)
+    val appDesc =
+      new ApplicationDescription(appName, maxCores, executorMemory, command, sparkHome, sc.ui.appUIAddress)
```
Comment (rxin) [inline core/src/main/scala/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala]: Mind wrapping the parameter line to make them fit in 100 characters wide?

i.e.

val appDesc = new ApplicationDescription(appName, maxCores, executorMemory, command, sparkHome,
  sc.ui.appUIAddress)

**Comment by rxin** [comment]
Thanks, this looks pretty good. I had two minor comments ...

File: core/src/main/scala/spark/storage/BlockManagerUI.scala (hunk 2 above)
Comment (JoshRosen) [inline core/src/main/scala/spark/storage/BlockManagerUI.scala:27]: I think this TODO was added by Matei in the the code that was moved here (see the lines below this in the diff).

**Comment by squito** [comment]
I fixed the line break.  As Josh said, I just copied the TODO from earlier.  I dunno why its a TODO (is there some alternate solution??) -- I could leave the comment and remove the "TODO".

**Comment by rxin** [comment]
Thanks, Imran.

**Pull request merged**